### PR TITLE
Clean channels with unexisting funding tx

### PR DESCRIPTION
**Pull request by pm47**
There are several separate but related changes in this PR:

(a) Fast close on scenarii where we have nothing at stake (instead of going to `CLOSING` state). The previous process was not only slower (we had to wait for confirmations), but it never resolved when the funding tx hadn't been confirmed. Note that there is still an edge case where the funding tx never gets confirmed, we are fundee and we have something at stake (`push_msat` > 0) (1).

(b) When *fundee*: after a certain timeout (2), if the funding tx hasn't reached `min_depth`, we cancel the channel.

(c) When *funder*: there is no timeout on the funding tx: however on restart, if we detect that our funding tx was doublespent (3), then we cancel the channel. Just because there is a doublespend doesn't mean that something malicious is going on: e.g. fee was to low, the tx was eventually removed from mempools and we just spent the inputs on something else).

Remaining questions:
(1) should we really consider that we have something at stake in that case?
(2) what should this timeout be?
(3) how do we implement this efficiently on electrum (for some reason the issue of non-published tx mostly happens with this backend)

### CODE REVIEW

**Comment by pm47** [comment]
@stone please advise on (3)

**Comment by sstone** [comment]
I don't know yet how to properly handle (3) and have created a specific issue.

**Comment by pm47** [comment]
@sstone I don't think that works since `ElectrumWallet` is stateless: channels will ask for double spends at startup, when the wallet is still uninitialized and it will always answer "no".

**Comment by pm47** [comment]
With the latest merge from master (which includes #873) I think it works. Even if electrum wallet doesn't see the double spend the first time, then the state will be stored and next time it won't miss it.

In other words it is ok to have some false positives occasionnally, what we really don't want is false positives.

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Channel.scala [hunk 1]
```diff
@@ -65,6 +66,9 @@ object Channel {
   // since BOLT 1.1, there is a max value for the refund delay of the main commitment tx
   val MAX_TO_SELF_DELAY = 2016
 
+  // as a fundee, we will wait that much time for the funding tx to confirm (funder will rely on the funding tx being double-spent)
+  val FUNDING_TIMEOUT_FUNDEE = 5 days
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Channel.scala]: New constant: 5 days

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Channel.scala [hunk 2]
```diff
@@ -202,6 +209,32 @@ class Channel(val nodeParams: NodeParams, wallet: EclairWallet, remoteNodeId: Pu
...
+                  case Success(false) => ()
+                  case Failure(t) => log.error(t, s"error while testing status of funding tx fundingTxid=${fundingTx.txid}: ")
+                }
+              case _ => ()
+            }
+          } else { // FUNDEE
+            // this is a bit tricky: let's say we shut down eclair right after someone opened a channel to us, and didn't start it up before a very long time
+            // we don't want the timeout to expire right away, because the watcher could be syncing or be busy, and may only notice the funding tx after some time
+            // so we always give us 10 minutes before doing anything
+            val delay = Funding.computeFundingTimeout(Platform.currentTime / 1000, funding.waitingSince, delay = FUNDING_TIMEOUT_FUNDEE, minDelay = 10 minutes)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Channel.scala]: New constant: 10 minutes

**Comment by pm47** [comment]
The fast close causes a regression with dataloss protection (see #854), because if we have nothing at stake we won't publish anything in case of error (even if our peer asks us to).

File: eclair-core/src/main/scala/fr/acinq/eclair/wire/ChannelCodecs.scala [hunk 3]
```diff
@@ -221,8 +221,18 @@ object ChannelCodecs extends Logging {
...
+    ("commitments" | commitmentsCodec) ::
+      ("fundingTx" | provide[Option[Transaction]](None)) ::
+      ("waitingSince" | provide(compat.Platform.currentTime / 1000)) ::
+      ("deferred" | optional(bool, fundingLockedCodec)) ::
+      ("lastSent" | either(bool, fundingCreatedCodec, fundingSignedCodec))).as[DATA_WAIT_FOR_FUNDING_CONFIRMED].decodeOnly
+
   val DATA_WAIT_FOR_FUNDING_CONFIRMED_Codec: Codec[DATA_WAIT_FOR_FUNDING_CONFIRMED] = (
     ("commitments" | commitmentsCodec) ::
+      ("fundingTx" | optional(bool, txCodec)) ::
+      ("timestamp" | int64) ::
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/wire/ChannelCodecs.scala]:
```suggestion
      ("waitingSince" | int64) ::
```

**Pull request merged**